### ISSUE TITLE
Use the sneaky-throws pattern for propagating checked exceptions instead of wrapping in runtime exceptions

### ISSUE THREAD

**Issue by gunnarmorling** [mapstruct/mapstruct#570]

**Comment by sjaakd** [comment]
I'm not sure for the following reasons:
1. We always put 'readable' code first. I don't think this kind of hack helps understand  the code, since the reader has to understand type-erasure (which is a tricky mechanism anyway in java).
2. I use a dedicated `TransformException` for all transformations (hand-written & generated). So I want to make sure that the caller of any method on a `Mapper` needs to catch the `TransformException` by contract.
3. As @agudian  puts it: its a bit hacky.
4. I seem to remember that I read somewhere that there was an issue with exceptions and generics. Of course.. when you need it you never can find it again ;)

**Comment by gunnarmorling** [comment]
> We always put 'readable' code first.

I agree it's important, but runtime behavior of the generated code is even more important. All users will *run* the generated code, only some will *read* it. Hence the way I see the priorities.

Wrapping all checked exceptions declared by used mapping methods into runtime exceptions makes live of MapStruct users harder: They either need to unwrap the runtime exception or will display/log longer stack traces than needed.

In a way I consider the wrapping into a runtime exception a "hack" we just need to do in order to deal with the annoyance that checked exceptions in Java are. Admittedly, the sneaky-throws approach is hacky, but IMO it's the lesser evil.

> I use a dedicated TransformException for all transformations (hand-written & generated)

Could you maybe show a short example how you use it? What's the gain of this pattern?

**Comment by sjaakd** [comment]
> Could you maybe show a short example how you use it?

Each declared mapping method should throw a TransformException.  TransformExceptions occur usually in in HandWritten mappers, e.g. transforming a geographic coordinate into JTS (java topology suite). But also to transform time / date formats into internal representation. Or data-blobs in XML. They are propagated through all the declared mapping methods to the caller of the mapping, so he can distinguish between transformation and exception occurring in a later phase (e.g. hibernate or other exceptions). 


> What's the gain of this pattern?

Well, our code makes a distinction between:
1. Validation (basically jSR 303 / Hibernate validation) , stuck on the BusinessDTO's
2. Tranformation e.g. from JAXB to BusinessDTO  and again to JPA entities. 

Both validation and transformation errors are mapped to a SOAP result (somehow caused by faulty user input), whereas a runtime exception is mapped to a SOAP Fault (is considered a system fault). So the distinction is important.

**Comment by sjaakd** [comment]
> Wrapping all checked exceptions declared by used mapping methods into runtime exceptions makes live of MapStruct users harder: They either need to unwrap the runtime exception or will display/log longer stack traces than needed.

You mean  undeclared exceptions right?. I think the declared ones are handled gracefully now..

**Comment by gunnarmorling** [comment]
Yes, right. Sorry for not being clearer. This only affects checked exceptions raised by methods invoked by generated methods and not declared by the latter themselves. These we wrap into `RuntimeException` atm. And that's what I seek to avoid with that proposal.

All other exceptions (unchecked exceptions and checked exceptions declared by mapping methods themselves) should be re-thrown as is.

**Comment by sjaakd** [comment]
> All other exceptions (unchecked exceptions and checked exceptions declared by mapping methods themselves) should be re-thrown as is.

Ahhh... Then I withdraw my objections..(Your honor) :smile:

**Comment by gunnarmorling** [comment]
Cool :) I'll see what I can do.
